pkg/service/service.go

```diff
@@ -160,13 +195,122 @@ func NewService(monitorNotify monitorNotify) *Service {
 		backendRefCount: counter.StringCounter{},
 		backendByHash:   map[string]*lb.Backend{},
 		monitorNotify:   monitorNotify,
+		envoyCache:      envoyCache,
 		healthServer:    localHealthServer,
 		lbmap:           lbmap.New(maglev, maglevTableSize),
+		l7lbSvcs:        map[string]*L7LBInfo{},
 	}
 	svc.lastUpdatedTs.Store(time.Now())
+
 	return svc
 }
 
+// RegisterL7LBService makes the given service to be locally forwarded to the
+// given proxy port, if non-zero, or for backend sync only, if zero.
+func (s *Service) RegisterL7LBService(name, namespace string, cecName string, proxyPort uint16) error {
```

Comment:
I had a bit of back and forth with the changes in `pkg/service`. I was wondering whether it would not be possible for the envoy config watcher to call the `UpsertService()` method, and move all envoy related bits from the `pkg/service` to the watcher or so?

My main concern is that we are breaking the separation of concerns wrt the service handling. The intention of the `pkg/service` is just to update the lbmap BPF maps. With this PR we are leaking part of envoy service management logic into the package.

I wonder why e.g. `upsertEnvoyEndpoints()` cannot be invoked from the envoy-related watcher?

Reply:
CiliumEnvoyConfig watcher does not know when endpoints are added/removed, but I managed to move all this Envoy specific code to `pkg/envoy` by importing `pkg/service` from there. I also exported `svcInfo` (as `SvcInfo`), and the three fields of it needed for generating the Envoy endpoints.

Reply:
I have the following suggestions:
1) Invoke the registration of service endpoints via `pkg/envoy` in https://github.com/cilium/cilium/blob/1381542cf12c1438c2f458cc4db4e83455c30c78/pkg/k8s/watchers/watcher.go#L822 (in this place we know endpoints).
2) Replace `RegisterL7LBService()` with `SetL7LB()`, and call it from `pkg/k8s/watchers/cilium_envoy_config.go` instead.

These would allow to eliminate any Envoy related bits from pkg/service, and thus would address my concern:
> My main concern is that we are breaking the separation of concerns wrt the service handling. The intention of the pkg/service is just to update the lbmap BPF maps. With this PR we are leaking part of envoy service management logic into the package.

Reply:
I already addressed this concern as I indicated above:
> CiliumEnvoyConfig watcher does not know when endpoints are added/removed, but I managed to move all this Envoy specific code to pkg/envoy by importing pkg/service from there.

There is no dependency from `pkg/service` to Envoy any more:
```
service % git grep "pkg/envoy"
service % 
```

Reply:
> There is no dependency from pkg/service to Envoy any more:

I'm talking about `envoyCache` and friends in `pkg/service/service.go`.

Reply:
Could you elaborate a bit on your suggestion above. I think `pkg/service` needs to be the source of the "backend changed trigger", are you saying `pkg/service` should call to `pkg/k8s/watchers` whenever any backend of any service is added/updated/removed?

Currently `pkg/service` is not importing anything from `pkg/k8s`, adding that might create an import loop.

Reply:
So my logic was to add enough state to `pkg/service` to be able to trigger envoy endpoint updates when needed, using an interface to avoid direct dependency.

Would adding a new file in `pkg/service` with all (or most) of the new state address your concern?

Reply:
Or are you saying that `pkg/k8s/watchers` intrinsically knows all service backend adds/updates/deletes?

Reply:
> Or are you saying that pkg/k8s/watchers intrinsically knows all service backend adds/updates/deletes?

Discussed offline. Just for others - yes, watchers know service backends.